dict: Compilation should check string literal

pranithk opened the pull request:
Recently I made a mistake where a variable is passed instead of string
literal to _sizen variant but the compilation succeeded. This can lead
to serious bugs. Added guard rails to prevent this from happening in
future.

fixes: #2410
Change-Id: I8cf1408b4a0a497903162de32e2cfb008ad7853c
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
/run regression

pranithk commented:
will raise issue in the morning

pranithk commented:
/run regression

pranithk commented:
@Shwetha-Acharya Noticing that geo-rep .ts are timing out nowadays.
Examples:
https://build.gluster.org/job/gh_centos7-regression/1142/console
https://build.gluster.org/job/gh_centos7-regression/1191/console

Could you check if there is any issue with geo-rep?

pranithk commented:
/run regression

Shwetha-Acharya commented:
@pranithk I do not see any issue with geo-rep. There are 

> https://build.gluster.org/job/gh_centos7-regression/1142/console
why is that code is of before the sensitive language patch? I still see master and slave in logs here.

>https://build.gluster.org/job/gh_centos7-regression/1191/console

[2021-05-07 02:02:22.982703] E [primary(worker /d/backends/primary5):789:log_failures] _GPrimary: ENTRY FAILED [{data=({'uid': 0, 'skip_entry': False, 'gfid': '7e9bc53b-4e53-4366-bb2d-57778ec10be8', 'gid': 0, 'mode': 16877, 'entry': '.gfid/00000000-0000-0000-0000-000000000001/history_d3', 'op': 'MKDIR'}, 2, {'secondary_isdir': False, 'dst': False, 'secondary_gfid': None, 'secondary_name': None, 'gfid_mismatch': False, 'name_mismatch': False})}]
[2021-05-07 02:02:22.982886] E [syncdutils(worker /d/backends/primary5):320:log_raise_exception] <top>: The above directory failed to sync. Please fix it to proceed further.

There is repeated entry fails, even though the gfid mismatch is resolved. 
Chances are that RPC communication between the primary gsyncd module and secondary gsyncd module is broken, there could be some transient ssh or fuse related issues in the machines.

I see that regression is re run. Lets see if it is any transient issue.

pranithk commented:
> @pranithk I do not see any issue with geo-rep. There are
> 
> > https://build.gluster.org/job/gh_centos7-regression/1142/console
> > why is that code is of before the sensitive language patch? I still see master and slave in logs here.
> 
> > https://build.gluster.org/job/gh_centos7-regression/1191/console
> 
> [2021-05-07 02:02:22.982703] E [primary(worker /d/backends/primary5):789:log_failures] _GPrimary: ENTRY FAILED [{data=({'uid': 0, 'skip_entry': False, 'gfid': '7e9bc53b-4e53-4366-bb2d-57778ec10be8', 'gid': 0, 'mode': 16877, 'entry': '.gfid/00000000-0000-0000-0000-000000000001/history_d3', 'op': 'MKDIR'}, 2, {'secondary_isdir': False, 'dst': False, 'secondary_gfid': None, 'secondary_name': None, 'gfid_mismatch': False, 'name_mismatch': False})}]
> [2021-05-07 02:02:22.982886] E [syncdutils(worker /d/backends/primary5):320:log_raise_exception] : The above directory failed to sync. Please fix it to proceed further.
> 
> There is repeated entry fails, even though the gfid mismatch is resolved.
> Chances are that RPC communication between the primary gsyncd module and secondary gsyncd module is broken, there could be some transient ssh or fuse related issues in the machines.
> 
> I see that regression is re run. Lets see if it is any transient issue.

Thanks for looking into it! Appreciate it!

pranithk commented:
/run regression

pranithk commented:
/run regression

mykaul commented on libglusterfs/src/glusterfs/dict.h:
```diff
@@ -21,31 +21,38 @@ typedef struct _data data_t;
 typedef struct _data_pair data_pair_t;
 
-#define dict_set_sizen(this, key, value) dict_setn(this, key, SLEN(key), value)
+#define DICT_STR_MUST_BE_LITERAL(__x) "" __x
```
Question: does it make sense, instead, change the definition of SLEN() macro to include this empty string?

pranithk replied:
On my machine that gives a warning, not an error:
"Suspicious usage of 'sizeof(char*)'; do you mean 'strlen'?"

pranithk replied:
Wait, gcc seems to be giving error, but clang is giving warning, I think. Let me check a bit more and get back

pranithk replied:
I think it works, let me change it and send the patch

pranithk replied:
Looks like it found a bug:
```
bit-rot-scrub.c: In function ‘br_scrubber_fetch_option’:
bit-rot-scrub.c:1399:26: error: expected ‘)’ before ‘opt’
 1399 |         GF_OPTION_RECONF(opt, *value, options, str, error_return);
      |                          ^~~
../../../../../libglusterfs/src/glusterfs/glusterfs.h:84:45: note: in definition of macro ‘GLUSTER_STR_MUST_BE_LITERAL’
   84 | #define GLUSTER_STR_MUST_BE_LITERAL(__x) "" __x
      |                                             ^~~
../../../../../libglusterfs/src/glusterfs/options.h:323:57: note: in expansion of macro ‘SLEN’
  323 |         if (xlator_option_reconf_##type(THIS, opt, key, SLEN(key), &(val)))    \
      |                                                         ^~~~
bit-rot-scrub.c:1399:9: note: in expansion of macro ‘GF_OPTION_RECONF’
 1399 |         GF_OPTION_RECONF(opt, *value, options, str, error_return);
      |         ^~~~~~~~~~~~~~~~
../../../../../libglusterfs/src/glusterfs/glusterfs.h:85:26: note: to match this ‘(’
   85 | #define SLEN(str) (sizeof(GLUSTER_STR_MUST_BE_LITERAL(str)) - 1)
      |                          ^
../../../../../libglusterfs/src/glusterfs/options.h:323:57: note: in expansion of macro ‘SLEN’
  323 |         if (xlator_option_reconf_##type(THIS, opt, key, SLEN(key), &(val)))    \
      |                                                         ^~~~
bit-rot-scrub.c:1399:9: note: in expansion of macro ‘GF_OPTION_RECONF’
 1399 |         GF_OPTION_RECONF(opt, *value, options, str, error_return);

```

pranithk replied:
It also found a valid usage that is not allowed now.
```
In file included from ../../../../libglusterfs/src/glusterfs/mem-pool.h:18,
                 from ../../../../libglusterfs/src/glusterfs/common-utils.h:48,
                 from glusterd-brick-ops.c:10:
glusterd-brick-ops.c: In function ‘glusterd_op_perform_add_bricks’:
glusterd-brick-ops.c:1050:26: error: expected ‘)’ before ‘brickinfo’
 1050 |                     SLEN(brickinfo->mount_dir));
      |                          ^~~~~~~~~
../../../../libglusterfs/src/glusterfs/glusterfs.h:84:45: note: in definition of macro ‘GLUSTER_STR_MUST_BE_LITERAL’
   84 | #define GLUSTER_STR_MUST_BE_LITERAL(__x) "" __x
      |                                             ^~~
glusterd-brick-ops.c:1050:21: note: in expansion of macro ‘SLEN’
 1050 |                     SLEN(brickinfo->mount_dir));
      |                     ^~~~
../../../../libglusterfs/src/glusterfs/glusterfs.h:85:26: note: to match this ‘(’
   85 | #define SLEN(str) (sizeof(GLUSTER_STR_MUST_BE_LITERAL(str)) - 1)
      |                          ^
glusterd-brick-ops.c:1050:21: note: in expansion of macro ‘SLEN’
 1050 |                     SLEN(brickinfo->mount_dir));
      |                     ^~~~
glusterd-brick-ops.c:1049:13: warning: ‘strncpy’ destination unchanged after copying no bytes [-Wstringop-truncation]
 1049 |             strncpy(brickinfo->mount_dir, brick_mount_dir,
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1050 |                     SLEN(brickinfo->mount_dir));
```

pranithk replied:
I personally think SLEN is a better place for it. I will let others also chime in with their opinions. If you also agree, I will add the following change to the diff and send the revised patch:
```
--- a/libglusterfs/src/glusterfs/options.h
+++ b/libglusterfs/src/glusterfs/options.h
@@ -320,7 +320,7 @@ DECLARE_RECONF_OPT(uint32_t, time);
 
 #define GF_OPTION_RECONF(key, val, opt, type, err_label)                       \
     do {                                                                       \
-        if (xlator_option_reconf_##type(THIS, opt, key, SLEN(key), &(val)))    \
+        if (xlator_option_reconf_##type(THIS, opt, key, strlen(key), &(val)))  \
             goto err_label;                                                    \
     } while (0)
 
diff --git a/xlators/mgmt/glusterd/src/glusterd-brick-ops.c b/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
index 84760cf47..bac356752 100644
--- a/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
+++ b/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
@@ -1047,7 +1047,7 @@ glusterd_op_perform_add_bricks(glusterd_volinfo_t *volinfo, int32_t count,
                 goto out;
             }
             strncpy(brickinfo->mount_dir, brick_mount_dir,
-                    SLEN(brickinfo->mount_dir));
+                    sizeof(brickinfo->mount_dir) - 1);
         }
 
         ret = glusterd_resolve_brick(brickinfo);
```

mykaul replied:
Yes, I've just noticed it myself while testing my proposal. Not only that it found a bug, but I am quite sure it's my bug, caused via 504f1f42a617e664875dd3dbb20c7cf54f08e5cd

pranithk replied:
Yeah, I feel there is more value with having it in SLEN, so re-sent the patch with all the changes I mentioned.

pranithk committed:
dict: Compilation should check string literal

Recently I made a mistake where a variable is passed instead of string
literal to _sizen variant but the compilation succeeded. This can lead
to serious bugs. Added guard rails to prevent this from happening in
future.

fixes: #2410
Change-Id: I8cf1408b4a0a497903162de32e2cfb008ad7853c
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
/run regression

gluster-ant commented:
4 test(s) needed retry 
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/ec/ec-background-heals.t
./tests/basic/fencing/afr-lock-heal-basic.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/glusterd-restart-shd-mux.t
